modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java

```diff
@@ -0,0 +1,330 @@
+                LOGGER.debug("moving database from [{}] to [{}]", databaseTmpFile, databaseFile);
+                Files.move(databaseTmpFile, databaseFile, StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING);
+                updateDatabase(databaseName, recordedMd5, databaseFile);
+                Files.delete(databaseTmpGzFile);
+            },
+            failure -> {
+                LOGGER.error((Supplier<?>) () -> new ParameterizedMessage("failed to download database [{}]", databaseName), failure);
+                try {
+                    Files.delete(databaseTmpFile);
+                    Files.delete(databaseTmpGzFile);
+                } catch (IOException ioe) {
+                    ioe.addSuppressed(failure);
+                    LOGGER.error("Unable to delete tmp database file after failure", ioe);
+                }
+            });
+    }
+
+    void updateDatabase(String databaseFileName, String recordedMd5, Path file) {
+        try {
+            LOGGER.info("database file changed [{}], reload database...", file);
+            DatabaseReaderLazyLoader loader = new DatabaseReaderLazyLoader(cache, file, recordedMd5);
+            DatabaseReaderLazyLoader existing = databases.put(databaseFileName, loader);
+            if (existing != null) {
+                existing.close();
+            }
+        } catch (Exception e) {
+            LOGGER.error((Supplier<?>) () -> new ParameterizedMessage("failed to update database [{}]", databaseFileName), e);
+        }
+    }
+
+    void removeStaleEntries(Collection<String> staleEntries) {
+        for (String staleEntry : staleEntries) {
+            try {
+                LOGGER.info("database [{}] no longer exists, cleaning up...", staleEntry);
+                DatabaseReaderLazyLoader existing = databases.remove(staleEntry);
+                assert existing != null;
+                existing.close(true);
+            } catch (Exception e) {
+                LOGGER.error((Supplier<?>) () -> new ParameterizedMessage("failed to clean database [{}]", staleEntry), e);
+            }
+        }
+    }
+
+    void retrieveDatabase(String databaseName,
+                          String expectedMd5,
+                          GeoIpTaskState.Metadata metadata,
+                          CheckedConsumer<byte[], IOException> chunkConsumer,
+                          CheckedRunnable<Exception> completedHandler,
+                          Consumer<Exception> failureHandler) {
+        // Need to run the search from a different thread, since this is executed from cluster state applier thread:
+        genericExecutor.accept(() -> {
+            MessageDigest md = MessageDigests.md5();
+            Long lastSortValue = null;
+
+            try {
+                do {
+                    SearchRequest searchRequest =
+                        createSearchRequest(databaseName, metadata.getFirstChunk(), metadata.getLastChunk(), lastSortValue);
```

Comment:
we can extract `metadata.getFirstChunk()` and `metadata.getLastChunk()` to local variables (even outside of the loop) and move this to single line